<h1><p style="text-align: center;">Pandas</p></h1>

![Pandas](https://upload.wikimedia.org/wikipedia/commons/thumb/e/ed/Pandas_logo.svg/512px-Pandas_logo.svg.png?20200209204934)

Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool,
built on top of the Python programming language.

## Table on Contents

* [What kind of data does pandas handle?](#01)
    * [Pandas data table representation](#01a)
    * [Each column in a DataFrame is a Series](#01b)    
    * [Do something with a DataFrame or Series](#01c)    
* [How do I read and write tabular data?](#02)    
    * [Interested in a technical summary of a DataFrame](#02a)    
    * [REMEMBER](#02b)    
* [Creating DataFrame from Scratch](#cdfs)
* [How do I select a subset of a DataFrame?](#03)    
    * [How do I select specific columns from a DataFrame?](#03a)    
    * [How do I filter specific rows from a DataFrame?](#03b)    
    * [How do I select specific rows and columns from a DataFrame?](#03c)    
    * [REMEMBER](#03d)    
* [How do I create plots in pandas?](#04)    
    * [REMEMBER](#04a)    
* [How to create new columns derived from existing columns?](#05)    
    * [REMEMBER](#05a)    
* [How to calculate summary statistics?](#06)
    * [Aggregating statistics](#06a)
    * [Aggregating statistics grouped by category](#06b)
    * [Count number of records by category](#06c)
    * [REMEMBER](#06d)

# What kind of data does pandas handle?<a class="anchor" id="01"></a>

### Pandas data table representation<a class="anchor" id="01a"></a>
![01_table_dataframe](https://pandas.pydata.org/docs/_images/01_table_dataframe.svg)

To manually store data in a table, create a DataFrame. When using a Python dictionary of lists, the dictionary keys will be used as column headers and the values in each list as columns of the DataFrame.

In [2]:
import pandas as pd

In [3]:
df = pd.DataFrame(
    {
        "Name": [
            "Braund, Mr. Owen Harris",
            "Allen, Mr. William Henry",
            "Bonnell, Miss. Elizabeth",
        ],
        "Age": [22, 35, 58],
        "Sex": ["male", "male", "female"],
    }
)

A DataFrame is a 2-dimensional data structure that can store data of different types (including characters, integers, floating point values, categorical data and more) in columns. It is similar to a spreadsheet, a SQL table or the ```data.frame``` in R.


- The table has 3 columns, each of them with a column label. The column labels are respectively ```Name```, ```Age``` and ```Sex```.
- The column ```Name``` consists of textual data with each value a string, the column ```Age``` are numbers and the column ```Sex``` is textual data.

In spreadsheet software, the table representation of our data would look very similar:

![01_table_spreadsheet](https://pandas.pydata.org/docs/_images/01_table_spreadsheet.png)


### Each column in a DataFrame is a Series<a class="anchor" id="01b"></a>
![01_table_series](https://pandas.pydata.org/docs/_images/01_table_series.svg)

Let's say you are just interested in working with the data in the column ```Age```

In [4]:
df["Age"]

0    22
1    35
2    58
Name: Age, dtype: int64

In [5]:
type(df["Age"])

pandas.core.series.Series

When selecting a single column of a pandas ```DataFrame```, the result is a pandas ```Series```. To select the column, use the column label in between square brackets ```[]```.

You can create a ```Series``` from scratch as well:

In [6]:
ages = pd.Series([22, 35, 58], name="Age")

In [7]:
ages

0    22
1    35
2    58
Name: Age, dtype: int64

In [8]:
type(ages)

pandas.core.series.Series

A pandas ```Series``` has no column labels, as it is just a single column of a ```DataFrame```. A ```Series``` does have row labels.



### Do something with a DataFrame or Series<a class="anchor" id="01c"></a>
Maximum Age of the passengers

We can do this on the DataFrame by selecting the Age column and applying ```max()```:

In [9]:
df["Age"].max()

58

Or to the Series:

In [10]:
ages.max()

58

As illustrated by the ```max()``` method, you can do things with a ```DataFrame``` or ```Series```. pandas provides a lot of functionalities, each of them a method you can apply to a ```DataFrame``` or ```Series```. As methods are functions, do not forget to use parentheses ```()```.

In [11]:
# interested in some basic statistics of the numerical data of my data table
df.describe()

Age
count   3.000000
mean   38.333333
std    18.230012
min    22.000000
25%    28.500000
50%    35.000000
75%    46.500000
max    58.000000

The ```describe()``` method provides a quick overview of the numerical data in a ```DataFrame```. As the ```Name``` and ```Sex``` columns are textual data, these are by default not taken into account by the ```describe()``` method.

Many pandas operations return a ```DataFrame``` or a ```Series```. The ```describe()``` method is an example of a pandas operation returning a pandas ```Series``` or a pandas ```DataFrame```.

----

# How do I read and write tabular data?<a class="anchor" id="02"></a>
![02_io_readwrite](https://pandas.pydata.org/docs/_images/02_io_readwrite.svg)

In [12]:
titanic = pd.read_csv("dataset/titanic.csv")

pandas provides the ```read_csv()``` function to read data stored as a csv file into a pandas ```DataFrame```. pandas supports many different file formats or data sources out of the box (csv, excel, sql, json, parquet, …), each of them with the prefix ```read_*```.

Make sure to always have a check on the data after reading in the data. When displaying a ```DataFrame```, the first and last 5 rows will be shown by default:

In [13]:
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [14]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

If you want to see the first 8 rows of a pandas DataFrame.

In [15]:
titanic.head(8)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S

To see the first N rows of a ```DataFrame```, use the ```head()``` method with the required number of rows (in this case 8) as argument.

Interested in the last N rows instead? pandas also provides a ```tail()``` method. For example, ```titanic.tail(10)``` will return the last 10 rows of the DataFrame.

A check on how pandas interpreted each of the column data types can be done by requesting the pandas ```dtypes``` attribute:

In [16]:
titanic.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

For each of the columns, the used data type is enlisted. The data types in this ```DataFrame``` are integers ```(int64)```, floats ```(float64)``` and strings ```(object)```.

When asking for the ```dtypes```, no brackets are used! ```dtypes``` is an attribute of a ```DataFrame``` and ```Series```. Attributes of ```DataFrame``` or ```Series``` do not need brackets. Attributes represent a characteristic of a ```DataFrame```/```Series```, whereas a method (which requires brackets) do something with the ```DataFrame```/```Series```

---

Saving the data to a spreadsheet

In [17]:
titanic.to_excel("dataset/titanic.xlsx", sheet_name="passengers", index=False)

Whereas ```read_*``` functions are used to read data to pandas, the ```to_*``` methods are used to store data. The ```to_excel()``` method stores the data as an excel file. In the example here, the sheet_name is named passengers instead of the default Sheet1. By setting ```index=False``` the row index labels are not saved in the spreadsheet.

The equivalent read function ```read_excel()``` will reload the data to a DataFrame:

In [18]:
# If you want to read the data that you just stored
# titanic = pd.read_excel("titanic.xlsx", sheet_name="passengers")
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### Interested in a technical summary of a DataFrame<a class="anchor" id="02a"></a>

In [19]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


The method ```info()``` provides technical information about a ```DataFrame```, so let’s explain the output in more detail:

- It is indeed a ```DataFrame```.
- There are 891 entries, i.e. 891 rows.
- Each row has a ```row``` label (aka the index) with values ranging from 0 to 890.
- The table has 12 columns. Most columns have a value for each of the ```rows``` (all 891 values are non-null). Some columns do have missing values and less than 891 non-null values.
- The columns ```Name```, ```Sex```, ```Cabin``` and Embarked consists of textual data (strings, aka ```object```). The other columns are numerical data with some of them whole numbers (aka ```integer```) and others are real numbers (aka ```float```).
- The kind of data (```characters```, ```integers```,…) in the different columns are summarized by listing the dtypes.
- The approximate amount of ```RAM``` used to hold the ```DataFrame``` is provided as well.

----

### REMEMBER<a class="anchor" id="02b"></a>

- Getting data in to pandas from many different file formats or data sources is supported by ```read_*``` functions.
- Exporting data out of pandas is provided by different ```to_*```methods.
- The ```head```/```tail```/```info``` methods and the ```dtypes``` attribute are convenient for a first check.

-----

# Creating DataFrame from Scratch<a class="anchor" id="cdfs"></a>

In [20]:
import numpy as np
np_array = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [21]:
# This is our array
np_array

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [22]:
# Let's reshape our array
np_array.reshape((3,3))

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [23]:
np_array = np_array.reshape((3,3))

In [24]:
np_array

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [25]:
# creating a list of index names
index_values = [1,2,3]
   
# creating a list of column names
column_values = ['Column_1', 'Column_2', 'Column_3']
  
# creating the dataframe
dummy_df = pd.DataFrame(data = np_array, 
                  index = index_values, 
                  columns = column_values)
  

In [26]:
# displaying the dataframe
dummy_df

Column_1  Column_2  Column_3
1         1         2         3
2         4         5         6
3         7         8         9

# How do I select a subset of a DataFrame?<a class="anchor" id="03"></a>

### How do I select specific columns from a DataFrame?<a class="anchor" id="03a"></a>
![03_subset_columns](https://pandas.pydata.org/docs/_images/03_subset_columns.svg)

In [27]:
# interested in the age of the Titanic passengers.
titanic["Age"]

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [28]:
ages = titanic["Age"]
ages.head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

To select a single column, use square brackets ```[]``` with the column name of the column of interest.

Each column in a ```DataFrame``` is a ```Series```. As a single column is selected, the returned object is a pandas ```Series```. We can verify this by checking the type of the output:

In [29]:
type(titanic["Age"])

pandas.core.series.Series

In [30]:
# And have a look at the shape of the output:
titanic["Age"].shape

(891,)

```DataFrame.shape``` is an attribute (remember tutorial on reading and writing, do not use parentheses for attributes) of a pandas ```Series``` and ```DataFrame``` containing the number of rows and columns: (nrows, ncolumns). A pandas ```Series``` is 1-dimensional and only the number of rows is returned.

In [31]:
# interested in the age and sex of the Titanic passengers.
age_sex = titanic[["Age", "Sex"]]

In [32]:
age_sex.head()

Age     Sex
0  22.0    male
1  38.0  female
2  26.0  female
3  35.0  female
4  35.0    male

To select multiple columns, use a list of column names within the selection brackets ```[]```.

Note:- The inner square brackets define a Python ```list``` with column names, whereas the outer brackets are used to select the data from a pandas ```DataFrame``` as seen in the previous example.

The returned data type is a pandas DataFrame:

In [33]:
type(titanic[["Age", "Sex"]])

pandas.core.frame.DataFrame

In [34]:
titanic[["Age", "Sex"]].shape

(891, 2)

The selection returned a ```DataFrame``` with 891 rows and 2 columns. Remember, a ```DataFrame``` is 2-dimensional with both a row and column dimension.

### How do I filter specific rows from a DataFrame?<a class="anchor" id="03b"></a>
![03_subset_rows](https://pandas.pydata.org/docs/_images/03_subset_rows.svg)

In [35]:
# interested in the passengers older than 35 years.
above_35 = titanic[titanic["Age"] > 35]

In [36]:
above_35.head()

PassengerId  Survived  Pclass  \
1             2         1       1   
6             7         0       1   
11           12         1       1   
13           14         0       3   
15           16         1       2   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   

    Parch    Ticket     Fare Cabin Embarked  
1       0  PC 17599  71.2833   C85        C  
6       0     17463  51.8625   E46        S  
11      0    113783  26.5500  C103        S  
13      5    347082  31.2750   NaN        S  
15      0    248706  16.0000   NaN        S

To select rows based on a conditional expression, use a condition inside the selection brackets ```[]```.

The condition inside the selection brackets ```titanic["Age"] > 35``` checks for which rows the Age column has a value larger than 35:

In [37]:
titanic["Age"] > 35

0      False
1       True
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Age, Length: 891, dtype: bool

The output of the conditional expression (```>```, but also ```==```, ```!=```, ```<```, ```<=```,… would work) is actually a pandas ```Series``` of boolean values (either ```True``` or ```False```) with the same number of rows as the original ```DataFrame```. Such a ```Series``` of boolean values can be used to filter the ```DataFrame``` by putting it in between the selection brackets ```[]```. Only rows for which the value is ```True``` will be selected.

We know from before that the original Titanic ```DataFrame``` consists of 891 rows. Let’s have a look at the number of rows which satisfy the condition by checking the ```shape``` attribute of the resulting ```DataFrame``` above_35:

In [38]:
above_35.shape

(217, 12)

### How do I select specific rows and columns from a DataFrame?<a class="anchor" id="03c"></a>
![03_subset_columns_rows](https://pandas.pydata.org/docs/_images/03_subset_columns_rows.svg)

In [39]:
# interested in the names of the passengers older than 35 years.
adult_names = titanic.loc[titanic["Age"] > 35, "Name"]

In [80]:
adult_names

1      Cumings, Mrs. John Bradley (Florence Briggs Th...
6                                McCarthy, Mr. Timothy J
11                              Bonnell, Miss. Elizabeth
13                           Andersson, Mr. Anders Johan
15                      Hewlett, Mrs. (Mary D Kingcome) 
                             ...                        
865                             Bystrom, Mrs. (Karolina)
871     Beckwith, Mrs. Richard Leonard (Sallie Monypeny)
873                          Vander Cruyssen, Mr. Victor
879        Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)
885                 Rice, Mrs. William (Margaret Norton)
Name: Name, Length: 217, dtype: object

### df.iloc & df.loc

In this case, a subset of both rows and columns is made in one go and just using selection brackets ```[]``` is not sufficient anymore. The ```loc```/```iloc``` operators are required in front of the selection brackets []. When using ```loc```/```iloc```, the part before the comma is the rows you want, and the part after the comma is the columns you want to select.

When using the column names, row labels or a condition expression, use the ```loc``` operator in front of the selection brackets ```[]```. For both the part before and after the comma, you can use a single label, a list of labels, a slice of labels, a conditional expression or a colon. Using a colon specifies you want to select all rows or columns.

In [41]:
# interested in rows 10 till 25 and columns 3 to 5.
titanic.iloc[9:25, 2:5]

Pclass                                               Name     Sex
9        2                Nasser, Mrs. Nicholas (Adele Achem)  female
10       3                    Sandstrom, Miss. Marguerite Rut  female
11       1                           Bonnell, Miss. Elizabeth  female
12       3                     Saundercock, Mr. William Henry    male
13       3                        Andersson, Mr. Anders Johan    male
14       3               Vestrom, Miss. Hulda Amanda Adolfina  female
15       2                   Hewlett, Mrs. (Mary D Kingcome)   female
16       3                               Rice, Master. Eugene    male
17       2                       Williams, Mr. Charles Eugene    male
18       3  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female
19       3                            Masselmani, Mrs. Fatima  female
20       2                               Fynney, Mr. Joseph J    male
21       2                              Beesley, Mr. Lawrence    male
22       3                        McGowan, Miss. Anna "Annie"  female
23       1                       Sloper, Mr. William Thompson    male
24       3                      Palsson, Miss. Torborg Danira  female

Again, a subset of both rows and columns is made in one go and just using selection brackets ```[]``` is not sufficient anymore. When specifically interested in certain rows and/or columns based on their position in the table, use the ```iloc``` operator in front of the selection brackets ```[]```.

When selecting specific rows and/or columns with ```loc``` or ```iloc```, new values can be assigned to the selected data. For example, to assign the name anonymous to the first 3 elements of the third column:

In [42]:
titanic.iloc[0:3, 3] = "anonymous"

In [43]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                                     anonymous    male  22.0      1      0   
1                                     anonymous  female  38.0      1      0   
2                                     anonymous  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S

### Let's say we want the Titanic passengers from cabin class 2

In [44]:
class_2 = titanic[(titanic["Pclass"] == 2)]

In [45]:
class_2.head()

PassengerId  Survived  Pclass                                 Name  \
9            10         1       2  Nasser, Mrs. Nicholas (Adele Achem)   
15           16         1       2     Hewlett, Mrs. (Mary D Kingcome)    
17           18         1       2         Williams, Mr. Charles Eugene   
20           21         0       2                 Fynney, Mr. Joseph J   
21           22         1       2                Beesley, Mr. Lawrence   

       Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
9   female  14.0      1      0  237736  30.0708   NaN        C  
15  female  55.0      0      0  248706  16.0000   NaN        S  
17    male   NaN      0      0  244373  13.0000   NaN        S  
20    male  35.0      0      0  239865  26.0000   NaN        S  
21    male  34.0      0      0  248698  13.0000   D56        S

### Let's say we want the Titanic passengers from cabin class 2 and 3.

In [46]:
class_23 = titanic[(titanic["Pclass"] == 2) | (titanic["Pclass"] == 3)]

In [47]:
class_23.head()

PassengerId  Survived  Pclass                            Name     Sex  \
0            1         0       3                       anonymous    male   
2            3         1       3                       anonymous  female   
4            5         0       3        Allen, Mr. William Henry    male   
5            6         0       3                Moran, Mr. James    male   
7            8         0       3  Palsson, Master. Gosta Leonard    male   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0  22.0      1      0         A/5 21171   7.2500   NaN        S  
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4  35.0      0      0            373450   8.0500   NaN        S  
5   NaN      0      0            330877   8.4583   NaN        Q  
7   2.0      3      1            349909  21.0750   NaN        S

Note

When combining multiple conditional statements, each condition must be surrounded by parentheses ```()```. Moreover, you can not use ```or```/```and``` but need to use the ```or``` operator ```|``` and the ```and``` operator ```&```.

----


Similar to the conditional expression, the ```isin()``` conditional function returns a ```True``` for each row the values are in the provided list. To filter the rows based on such a function, use the conditional function inside the selection brackets ```[]```. In this case, the condition inside the selection brackets ```titanic["Pclass"].isin([2, 3])``` checks for which rows the Pclass column is either 2 or 3.

The above is equivalent to filtering by rows for which the class is either 2 or 3 and combining the two statements with an ```|``` (```or```) operator:

### Let's say we want the Titanic passengers from cabin class 2.

In [48]:
class_2 = titanic[titanic["Pclass"].isin([2])]

In [49]:
class_2.head()

PassengerId  Survived  Pclass                                 Name  \
9            10         1       2  Nasser, Mrs. Nicholas (Adele Achem)   
15           16         1       2     Hewlett, Mrs. (Mary D Kingcome)    
17           18         1       2         Williams, Mr. Charles Eugene   
20           21         0       2                 Fynney, Mr. Joseph J   
21           22         1       2                Beesley, Mr. Lawrence   

       Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
9   female  14.0      1      0  237736  30.0708   NaN        C  
15  female  55.0      0      0  248706  16.0000   NaN        S  
17    male   NaN      0      0  244373  13.0000   NaN        S  
20    male  35.0      0      0  239865  26.0000   NaN        S  
21    male  34.0      0      0  248698  13.0000   D56        S

### Let's say we want the Titanic passengers from cabin class 2 and 3.

In [50]:
class_23 = titanic[titanic["Pclass"].isin([2,3])]

In [51]:
class_23.head()

PassengerId  Survived  Pclass                            Name     Sex  \
0            1         0       3                       anonymous    male   
2            3         1       3                       anonymous  female   
4            5         0       3        Allen, Mr. William Henry    male   
5            6         0       3                Moran, Mr. James    male   
7            8         0       3  Palsson, Master. Gosta Leonard    male   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0  22.0      1      0         A/5 21171   7.2500   NaN        S  
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4  35.0      0      0            373450   8.0500   NaN        S  
5   NaN      0      0            330877   8.4583   NaN        Q  
7   2.0      3      1            349909  21.0750   NaN        S

### Want to work with passenger data for which the age is known.

Pandas ```DataFrame.isna()``` or ```DataFrame.isnull()``` method returns a boolean mask where ```True``` is set for missing values (NaN) and ```False``` for non-missing values.

In [52]:
titanic["Age"].isna()

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888     True
889    False
890    False
Name: Age, Length: 891, dtype: bool

In [53]:
titanic["Age"].isnull()

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888     True
889    False
890    False
Name: Age, Length: 891, dtype: bool

Opposite of this, we have ```DataFrame.notna()``` or ```DataFrame.notnull()```

In [54]:
titanic["Age"].notna()

0       True
1       True
2       True
3       True
4       True
       ...  
886     True
887     True
888    False
889     True
890     True
Name: Age, Length: 891, dtype: bool

In [55]:
titanic["Age"].notnull()

0       True
1       True
2       True
3       True
4       True
       ...  
886     True
887     True
888    False
889     True
890     True
Name: Age, Length: 891, dtype: bool

-----

In [56]:
titanic[titanic["Age"].notna()]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                       anonymous    male  22.0      1      0   
1                                       anonymous  female  38.0      1      0   
2                                       anonymous  female  26.0      0      0   
3    Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                        Allen, Mr. William Henry    male  35.0      0      0   
..                                            ...     ...   ...    ...    ...   
885          Rice, Mrs. William (Margaret Norton)  female  39.0      0      5   
886                         Montvila, Rev. Juozas    male  27.0      0      0   
887                  Graham, Miss. Margaret Edith  female  19.0      0      0   
889                         Behr, Mr. Karl Howell    male  26.0      0      0   
890                           Dooley, Mr. Patrick    male  32.0      0      0   

               Ticket     Fare Cabin Embarked  
0           A/5 21171   7.2500   NaN        S  
1            PC 17599  71.2833   C85        C  
2    STON/O2. 3101282   7.9250   NaN        S  
3              113803  53.1000  C123        S  
4              373450   8.0500   NaN        S  
..                ...      ...   ...      ...  
885            382652  29.1250   NaN        Q  
886            211536  13.0000   NaN        S  
887            112053  30.0000   B42        S  
889            111369  30.0000  C148        C  
890            370376   7.7500   NaN        Q  

[714 rows x 12 columns]

In [57]:
age_no_na = titanic[titanic["Age"].notna()]

The ```notna()``` conditional function returns a ```True``` for each row the values are not a ```Null``` value. As such, this can be combined with the selection brackets ```[]``` to filter the data table.

You might wonder what actually changed, as the first 5 lines are still the same values. One way to verify is to check if the ```shape``` has changed:

In [58]:
age_no_na.shape

(714, 12)

### REMEMBER<a class="anchor" id="03d"></a>
- When selecting subsets of data, square brackets ```[]``` are used.
- Inside these brackets, you can use a single column/row label, a list of column/row labels, a slice of labels, a conditional expression or a colon.
- Select specific rows and/or columns using ```loc``` when using the row and column names.
- Select specific rows and/or columns using ```iloc``` when using the positions in the table.
- You can assign new values to a selection based on ```loc```/```iloc```.

-----



# How to create new columns derived from existing columns?<a class="anchor" id="05"></a>
![05_newcolumn_1](https://pandas.pydata.org/docs/_images/05_newcolumn_1.svg)

In [59]:
air_quality = pd.read_csv("dataset/air_quality_no2.csv", index_col=0, parse_dates=True)
air_quality.head()

station_antwerp  station_paris  station_london
datetime                                                           
2019-05-07 02:00:00              NaN            NaN            23.0
2019-05-07 03:00:00             50.5           25.0            19.0
2019-05-07 04:00:00             45.0           27.7            19.0
2019-05-07 05:00:00              NaN           50.4            16.0
2019-05-07 06:00:00              NaN           61.9             NaN

In [60]:
# Want to express the NO2 concentration of the station in London in mg/m.
air_quality["station_london"] * 1.882

datetime
2019-05-07 02:00:00    43.286
2019-05-07 03:00:00    35.758
2019-05-07 04:00:00    35.758
2019-05-07 05:00:00    30.112
2019-05-07 06:00:00       NaN
                        ...  
2019-06-20 22:00:00       NaN
2019-06-20 23:00:00       NaN
2019-06-21 00:00:00       NaN
2019-06-21 01:00:00       NaN
2019-06-21 02:00:00       NaN
Name: station_london, Length: 1035, dtype: float64

In [61]:
air_quality["london_mg_per_cubic"] = air_quality["station_london"] * 1.882

To create a new column, use the ```[]``` brackets with the new column name at the left side of the assignment.

The calculation of the values is done element-wise. This means all values in the given column are multiplied by the value 1.882 at once. You do not need to use a loop to iterate each of the rows!

![05_newcolumn_2](https://pandas.pydata.org/docs/_images/05_newcolumn_2.svg)

In [62]:
# check the ratio of the values in Paris versus Antwerp and save the result in a new column.
air_quality["station_paris"] / air_quality["station_antwerp"]

datetime
2019-05-07 02:00:00         NaN
2019-05-07 03:00:00    0.495050
2019-05-07 04:00:00    0.615556
2019-05-07 05:00:00         NaN
2019-05-07 06:00:00         NaN
                         ...   
2019-06-20 22:00:00         NaN
2019-06-20 23:00:00         NaN
2019-06-21 00:00:00         NaN
2019-06-21 01:00:00         NaN
2019-06-21 02:00:00         NaN
Length: 1035, dtype: float64

In [63]:
air_quality["ratio_paris_antwerp"] = (air_quality["station_paris"] / air_quality["station_antwerp"])

The calculation is again element-wise, so the ```/``` is applied for the values in each row.

Also other mathematical operators (```+```, ```-```, ```*```, ```/```,…) or logical operators (```<```, ```>```, ```==```,…) work element-wise. The latter was already used in the subset data tutorial to filter rows of a table using a conditional expression.

If you need more advanced logic, you can use arbitrary Python code via ```apply()```(will cover later).

In [64]:
# rename the data columns to the corresponding station identifiers used by OpenAQ.
air_quality.rename(columns={"station_antwerp": "BETR801","station_paris": "FR04014","station_london": "London Westminster",})

BETR801  FR04014  London Westminster  \
datetime                                                    
2019-05-07 02:00:00      NaN      NaN                23.0   
2019-05-07 03:00:00     50.5     25.0                19.0   
2019-05-07 04:00:00     45.0     27.7                19.0   
2019-05-07 05:00:00      NaN     50.4                16.0   
2019-05-07 06:00:00      NaN     61.9                 NaN   
...                      ...      ...                 ...   
2019-06-20 22:00:00      NaN     21.4                 NaN   
2019-06-20 23:00:00      NaN     24.9                 NaN   
2019-06-21 00:00:00      NaN     26.5                 NaN   
2019-06-21 01:00:00      NaN     21.8                 NaN   
2019-06-21 02:00:00      NaN     20.0                 NaN   

                     london_mg_per_cubic  ratio_paris_antwerp  
datetime                                                       
2019-05-07 02:00:00               43.286                  NaN  
2019-05-07 03:00:00               35.758             0.495050  
2019-05-07 04:00:00               35.758             0.615556  
2019-05-07 05:00:00               30.112                  NaN  
2019-05-07 06:00:00                  NaN                  NaN  
...                                  ...                  ...  
2019-06-20 22:00:00                  NaN                  NaN  
2019-06-20 23:00:00                  NaN                  NaN  
2019-06-21 00:00:00                  NaN                  NaN  
2019-06-21 01:00:00                  NaN                  NaN  
2019-06-21 02:00:00                  NaN                  NaN  

[1035 rows x 5 columns]

In [65]:
air_quality_renamed = air_quality.rename(columns={"station_antwerp": "BETR801","station_paris": "FR04014","station_london": "London Westminster",})

In [66]:
air_quality_renamed.head()

BETR801  FR04014  London Westminster  \
datetime                                                    
2019-05-07 02:00:00      NaN      NaN                23.0   
2019-05-07 03:00:00     50.5     25.0                19.0   
2019-05-07 04:00:00     45.0     27.7                19.0   
2019-05-07 05:00:00      NaN     50.4                16.0   
2019-05-07 06:00:00      NaN     61.9                 NaN   

                     london_mg_per_cubic  ratio_paris_antwerp  
datetime                                                       
2019-05-07 02:00:00               43.286                  NaN  
2019-05-07 03:00:00               35.758             0.495050  
2019-05-07 04:00:00               35.758             0.615556  
2019-05-07 05:00:00               30.112                  NaN  
2019-05-07 06:00:00                  NaN                  NaN

The ```rename()``` function can be used for both row labels and column labels. Provide a dictionary with the keys the current names and the values the new names to update the corresponding names.

The mapping should not be restricted to fixed names only, but can be a mapping function as well. For example, converting the column names to lowercase letters can be done using a function as well:

In [67]:
air_quality_renamed = air_quality_renamed.rename(columns=str.lower)

In [68]:
air_quality_renamed.head()

betr801  fr04014  london westminster  \
datetime                                                    
2019-05-07 02:00:00      NaN      NaN                23.0   
2019-05-07 03:00:00     50.5     25.0                19.0   
2019-05-07 04:00:00     45.0     27.7                19.0   
2019-05-07 05:00:00      NaN     50.4                16.0   
2019-05-07 06:00:00      NaN     61.9                 NaN   

                     london_mg_per_cubic  ratio_paris_antwerp  
datetime                                                       
2019-05-07 02:00:00               43.286                  NaN  
2019-05-07 03:00:00               35.758             0.495050  
2019-05-07 04:00:00               35.758             0.615556  
2019-05-07 05:00:00               30.112                  NaN  
2019-05-07 06:00:00                  NaN                  NaN

### REMEMBER<a class="anchor" id="05a"></a>
- Create a new column by assigning the output to the ```DataFrame``` with a new column name in between the ```[]```.
- Operations are element-wise, no need to ```loop``` over rows.
- Use ```rename``` with a dictionary or function to rename row labels or column names.

-----

# How to calculate summary statistics?<a class="anchor" id="06"></a>

### Aggregating statistics<a class="anchor" id="06a"></a>
![06_aggregate](https://pandas.pydata.org/docs/_images/06_aggregate.svg)

In [69]:
# What is the average age of the Titanic passengers?
titanic["Age"].mean()

29.69911764705882

Different statistics are available and can be applied to columns with numerical data. Operations in general exclude missing data and operate across rows by default.

![06_reduction](https://pandas.pydata.org/docs/_images/06_reduction.svg)

In [70]:
# What is the median age and ticket fare price of the Titanic passengers?
titanic[["Age", "Fare"]].median()

Age     28.0000
Fare    14.4542
dtype: float64

The statistic applied to multiple columns of a ```DataFrame``` (the selection of two columns returns a DataFrame, see the subset data tutorial) is calculated for each numeric column.

The aggregating statistic can be calculated for multiple columns at the same time. Remember the describe function from the first tutorial?

In [71]:
titanic[["Age", "Fare"]].describe()

Age        Fare
count  714.000000  891.000000
mean    29.699118   32.204208
std     14.526497   49.693429
min      0.420000    0.000000
25%     20.125000    7.910400
50%     28.000000   14.454200
75%     38.000000   31.000000
max     80.000000  512.329200

Instead of the predefined statistics, specific combinations of aggregating statistics for given columns can be defined using the ```DataFrame.agg()``` method:

In [72]:
titanic.agg({"Age": ["min", "max", "median", "skew"],"Fare": ["min", "max", "median", "mean"],})

Age        Fare
min      0.420000    0.000000
max     80.000000  512.329200
median  28.000000   14.454200
skew     0.389108         NaN
mean          NaN   32.204208

### Aggregating statistics grouped by category<a class="anchor" id="06b"></a>

![06_groupby](https://pandas.pydata.org/docs/_images/06_groupby.svg)

In [73]:
# What is the average age for male versus female Titanic passengers?
titanic[["Sex", "Age"]].groupby("Sex").mean()

Age
Sex              
female  27.915709
male    30.726645

As our interest is the average age for each gender, a subselection on these two columns is made first: ```titanic[["Sex", "Age"]]```. Next, the ```groupby()``` method is applied on the ```Sex``` column to make a group per category. The average ```age``` for each ```gender``` is calculated and returned.

Calculating a given statistic (e.g. ```mean``` age) for each category in a column (e.g. male/female in the Sex column) is a common pattern. The ```groupby``` method is used to support this type of operations. This fits in the more general s```plit-apply-combine``` pattern:

- Split the data into groups
- Apply a function to each group independently
- Combine the results into a data structure

The apply and combine steps are typically done together in pandas.

In the previous example, we explicitly selected the 2 columns first. If not, the ```mean``` method is applied to each column containing numerical columns by ```passing numeric_only=True```:

In [74]:
titanic.groupby("Sex").mean(numeric_only=True)

PassengerId  Survived    Pclass        Age     SibSp     Parch  \
Sex                                                                      
female   431.028662  0.742038  2.159236  27.915709  0.694268  0.649682   
male     454.147314  0.188908  2.389948  30.726645  0.429809  0.235702   

             Fare  
Sex                
female  44.479818  
male    25.523893

It does not make much sense to get the average value of the ```Pclass```. If we are only interested in the average age for each ```gender```, the selection of columns (rectangular brackets ```[]``` as usual) is supported on the grouped data as well:

In [75]:
titanic.groupby("Sex")["Age"].mean()

Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64

![06_groupby_select_detail](https://pandas.pydata.org/docs/_images/06_groupby_select_detail.svg)

The ```Pclass``` column contains numerical data but actually represents 3 ```categories``` (or factors) with respectively the labels ‘1’, ‘2’ and ‘3’. Calculating statistics on these does not make much sense. Therefore, pandas provides a ```Categorical``` data type to handle this type of data. Will cover later.

In [76]:
# What is the mean ticket fare price for each of the sex and cabin class combinations?
titanic.groupby(["Sex", "Pclass"])["Fare"].mean()

Sex     Pclass
female  1         106.125798
        2          21.970121
        3          16.118810
male    1          67.226127
        2          19.741782
        3          12.661633
Name: Fare, dtype: float64

Grouping can be done by multiple columns at the same time. Provide the column names as a ```list``` to the ```groupby()``` method.

----

### Count number of records by category<a class="anchor" id="06c"></a>

![06_valuecounts](https://pandas.pydata.org/docs/_images/06_valuecounts.svg)

In [77]:
# What is the number of passengers in each of the cabin classes?
titanic["Pclass"].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

The ```value_counts()``` method counts the number of records for each category in a column.

The function is a shortcut, as it is actually a groupby operation in combination with counting of the number of records within each group:

In [78]:
titanic.groupby("Pclass")["Pclass"].count()

Pclass
1    216
2    184
3    491
Name: Pclass, dtype: int64

Note:- Both ```size``` and ```count``` can be used in combination with ```groupby```. Whereas ```size``` includes ```NaN``` values and just provides the number of rows (size of the table), ```count``` excludes the missing values. In the ```value_counts``` method, use the ```dropna``` argument to include or exclude the ```NaN``` values.

In [79]:
titanic["Pclass"].value_counts(dropna=False)

3    491
1    216
2    184
Name: Pclass, dtype: int64

### REMEMBER<a class="anchor" id="06d"></a>
- Aggregation statistics can be calculated on entire columns or rows.
- ```groupby``` provides the power of the split-apply-combine pattern.
- ```value_counts``` is a convenient shortcut to count the number of entries in each category of a variable.